util/circular_buffer.c

```diff
@@ -0,0 +1,228 @@
+	uint32_t	nb_elem;
+	/** Size of an user element */
+	uint32_t	elem_size;
+	/** Address to the buffer */
+	int8_t		*buff;
+	/** Pointer to reading location */
+	int8_t		*read_ptr;
+	/** Pointer to writing location */
+	int8_t		*write_ptr;
+	/** Set when the buffer is full */
+	bool		full;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+int32_t cb_init(struct circular_buffer **desc, uint32_t nb_elements,
+		uint32_t element_size)
+{
+	struct circular_buffer	*ldesc;
+	uint32_t		size;
+
+	if (!desc || !element_size || !nb_elements)
+		return FAILURE;
+
+	size = nb_elements * element_size;
+	if (size < nb_elements || size < element_size)
+		return FAILURE; //Integer overflow
+
+	ldesc = (struct circular_buffer*)calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		return FAILURE;
+
+	ldesc->buff = calloc(nb_elements, element_size);
+	if (!ldesc->buff) {
+		free(ldesc);
+		return FAILURE;
+	}
+
+	ldesc->size = size;
+	ldesc->elem_size = element_size;
+	ldesc->nb_elem = 0;
+	ldesc->read_ptr = ldesc->buff;
+	ldesc->write_ptr = ldesc->buff;
+
+	*desc = ldesc;
+
+	return SUCCESS;
+}
+
+int32_t cb_remove(struct circular_buffer *desc)
+{
+	if (!desc)
+		return FAILURE;
+
+	if (desc->buff)
+		free(desc->buff);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/* Convert pointer in buffer to offset from the beginning */
+static inline uint32_t ptr_to_offset(struct circular_buffer *desc, int8_t *ptr)
+{
+	return ptr - desc->buff;
+}
+
+/* Get size of the datt written to the buffer */
+static inline uint32_t get_unread_size(struct circular_buffer *desc)
+{
+	uint32_t		size;
+
+	if (desc->full)
+		return desc->size;
+
+	if (desc->write_ptr < desc->read_ptr) {
+		size = desc->size - ptr_to_offset(desc, desc->read_ptr);
+		size += desc->write_ptr - desc->buff;
+	} else {
+		size = desc->write_ptr - desc->read_ptr;
+	}
+
+	return size;
+}
+
+int32_t cb_size(struct circular_buffer *desc, uint32_t *nb_elements)
+{
+	if (!desc)
+		return FAILURE;
+
+	*nb_elements =  desc->nb_elem;
+
+	return SUCCESS;
+}
+
+int32_t cb_write(struct circular_buffer *desc, const void *data,
+		 uint32_t nb_elements)
+{
+	uint32_t		to_write_size;
+	uint32_t		size_to_end;
+
+	if (!desc || !nb_elements)
+		return FAILURE;
+
+	if (desc->full)
+		return FAILURE;
+
+	to_write_size = nb_elements * desc->elem_size;
+	/* Check for overflow */
+	if (to_write_size + (desc->nb_elem * desc->elem_size) > desc->size)
+		return FAILURE;
+	desc->nb_elem += nb_elements;
+
+	size_to_end = desc->size - ptr_to_offset(desc, desc->write_ptr);
+	if (to_write_size > size_to_end) {
+		memcpy(desc->write_ptr, data, size_to_end);
+		to_write_size -= size_to_end;
+		memcpy(desc->buff, (uint8_t *)data + size_to_end,
+		       to_write_size);
+		desc->write_ptr = desc->buff + to_write_size;
```

Comment:
What if the write would pass the read pointer? Do we lose that data or guard against that?